src/adapter/src/catalog/apply.rs

```diff
@@ -1092,6 +1024,113 @@ impl CatalogState {
     }
 }
 
+/// Most updates are applied one at a time, but during bootstrap, certain types are applied
+/// separately in a batch for performance reasons. A constraint is that updates must be applied in
+/// order. This method is modeled as a state machine that batches then applies groups of updates.
+enum BootstrapApplyState {
+    /// Additions of builtin views.
+    BuiltinViewAdditions(Vec<(&'static Builtin<NameReference>, GlobalId)>),
+    /// Item updates that aren't builtin view additions.
+    Items(Vec<StateUpdate>),
+    /// All other updates.
+    Updates(Vec<StateUpdate>),
+}
+
+impl BootstrapApplyState {
+    fn new(update: StateUpdate) -> BootstrapApplyState {
+        match update {
+            StateUpdate {
+                kind: StateUpdateKind::SystemObjectMapping(system_object_mapping),
+                diff: StateDiff::Addition,
+            } if matches!(
+                system_object_mapping.description.object_type,
+                CatalogItemType::View
+            ) =>
+            {
+                let view_addition = lookup_builtin_view_addition(system_object_mapping);
+                BootstrapApplyState::BuiltinViewAdditions(vec![view_addition])
+            }
+            StateUpdate {
+                kind: StateUpdateKind::IntrospectionSourceIndex(_),
+                ..
+            }
+            | StateUpdate {
+                kind: StateUpdateKind::SystemObjectMapping(_),
+                ..
+            }
+            | StateUpdate {
+                kind: StateUpdateKind::Item(_),
+                ..
+            } => BootstrapApplyState::Items(vec![update]),
+            update => BootstrapApplyState::Updates(vec![update]),
+        }
+    }
+
+    async fn apply(
+        &mut self,
+        state: &mut CatalogState,
+        retractions: &mut InProgressRetractions,
+    ) -> Vec<BuiltinTableUpdate<&'static BuiltinTable>> {
+        match self {
+            BootstrapApplyState::BuiltinViewAdditions(builtin_view_additions) => {
+                // We enable all "enable_for_item_parsing" feature flags when applying item updates
+                // during bootstrap.
+                //
+                // See [`CatalogState::with_enable_for_item_parsing`] for more details.
+                let restore = state.system_configuration.clone();
+                state.system_configuration.enable_for_item_parsing();
+                let builtin_table_updates =
+                    Catalog::parse_builtin_views(state, std::mem::take(builtin_view_additions))
```

Comment:
The doc-comments (https://doc.rust-lang.org/std/mem/fn.take.html) are as follows:

> Replaces `dest` with the default value of `T`, returning the previous `dest` value.

Essentially, it takes all of the elements out of a `Vec` leaving the `Vec` empty. I tried to have this method take `self` so we could just move the `Vec`s instead of calling `take`, but couldn't get the borrow checker to approve. I've had some ideas since I wrote this, so I'll try them out and see if it helps.

Reply:
Done